### ISSUE TITLE
Try Backdrop CMS on Platform.sh

### ISSUE THREAD

**Issue by serundeputy** [backdrop-ops/backdropcms.org#600]
@jenlampton or anyone who has an opinion.

from @crell

>We've made a number of changes to the way project templates work at Platform.sh recently, and I want to make sure Backdrop is updated accordingly.  In particular, we now have the ability to include an almost unlimited number of templates in the new-project-creation process.  We have about 50 now, and Backdrop should be added to that list.

>To that end, there's one of two courses of action to take.

>1) We retire the existing example repo.  (I'm not sure how much it's used; I see Jen merged something a few weeks ago but most of it is fairly out of date it looks like.)  Then Platform rebuilds it using our new template-builder build tool for a Platform-maintained template repository in the `platformsh-templates` GitHub org.

>2) Backdrop continues to maintain the template themselves, but updates it to current standards; the current config files are a bit out of date, and are using old and unsupported container versions right now.  (Eg, PHP 7.0.)  We likely want to then move the repository from our GitHub repo to Backdrop's.  Then we add a template definition file to our external templates repository (https://github.com/platformsh/templates-external/), so that it shows up in the new-project wizard.

>Both end up in the same place, but one makes it your responsibility to maintain and the other means you don't have as easy access to maintain it.

>In either case, it would also be supported by the new Deploy on Platform link/button we're about to release this month, which makes it easy to have a button on a site anywhere that creates a new Platform.sh project from an arbitrary template.  That means you would be able to have a "try Backdrop now on Platform.sh" button that gives someone a ready-to-roll Backdrop instance on Platform.sh, which you can include wherever you like.  A more formal announcement on that is coming soon-ish, but I wanted to loop you in beforehand so that we can get the template ready for it.

>Let me know how Backdrop would like to proceed.  Cheers.

**Comment by serundeputy** [comment]
I like option `1` but either way is fine with me.  It'd be nice to have a try Backdrop CMS on Platform.sh button on backdropcms.org! 

Probably either on their org landing page https://github.com/backdrop-ops/backdropcms.org/issues/598 and/or on the `/try-backdrop` page https://github.com/backdrop-ops/backdropcms.org/issues/555

**Comment by jenlampton** [comment]
It might be better to have platform create the repo, since they understand
the platform part best.

I don't have a problem with Platform owning the repo too, as long as it can
still get updated with every Backdrop release.

Since we understand Backdrop (and more importantly, our release timeline)
best, it might work out to have one of us file PRs for each update against
the platform repo, if that's still allowed. (We already have a release
checklist item to update Platform, well just change the instructions for
how to do it.)

If we wouldn't have the ability to help with updates if it were
Platform-owned, rather have the repo under Backdrop.

And I'd love to see that button on both pages :)

**Comment by Crell** [comment]
The caveat with our new build system is that it's not just a single repo; we actually use a Python script that pulls from an upstream Git source and then applies our files and patches to it.  That makes it much easier for us to safely manage 50+ templates since we know exactly what the changes are that we're making against upstream, but also therefore harder for 3rd parties to contribute.

Our build tool is here, for reference:

https://github.com/platformsh/template-builder

So the ability to help is still there, it's just a higher barrier to entry than it used to be.

**Comment by Crell** [comment]
(We're happy to advise/help with the Platform bits of the repo if you're maintaining it, to be clear.  In fact I'd prefer to be involved in the initial setup to make sure it's consistent with our conventions elsewhere.)

**Comment by jenlampton** [comment]
But if Platform is already pulling from an external repo, our updates might
be easier anyway. We'd just point to a new tag?

**Comment by Crell** [comment]
It depends on the complexity of the template.  For simple templates, such as Symfony, the process essentially boils down to:

* Git merge the latest tag from Symfony into our repo.
* Copy over a bunch of files into that directory, overwriting if necessary.
* Apply any patches we need to it.
* Composer update.
* Commit the result.
* Make new PR against our template repo.

That way, we can manually run "rebuild symfony" at any time and get the latest updates, without having to remember what exactly the changes we made were; the script makes them for us.  Naturally for more involved or less-PaaS-friendly tools the steps are the same but the details can get more involved.  Installers are the most common source of issue.

So once it's setup, if it's a well-behaved template it's just a matter of us running the script and merging the PR.  The latest-tag gets detected automatically.  If something needs to change in our add-ons, though, that requires a PR against our build tool.  (Which I'd totally accept from your team if you prefer to go that route.)

The alternative is you maintain a repo and do whatever the heck you want with it, however you want, and we just list it in the external templates index and call it a day.  Then what you do with that repo is your business.

**Comment by serundeputy** [comment]
I'd prefer to just maintain `/backdrop/backdrop` github repo and let p.sh pull in the tags and merge into your template repo.

**Comment by jenlampton** [comment]
I took a look at the Drupal 7 template, and I see that it's pointing at the
master branch, instead of a tag. I think for our audience a tag might be
safer, but I'm guessing that's possible too.

I saw that only modules themes and libraries are provided in the files
directory, but couldn't see the bit that determined where those get placed
within the docroot, or how else those might be managed.

I'd be most comfortable providing feedback on a template someone at
Platform set up for us, maybe based on what we've done in the current repo?

Let's start with having that live under under the platform org, and if we
run into any issues with that we can transfer it to Backdrop then.

**Comment by Crell** [comment]
The `files` directory is just rsync'ed over the checkout of the repo, so anything the upstream already provides that we don't need to touch, we don't touch.

One possible caveat is that, AFAIK, Backdrop like Drupal 7 wants to have the whole site in the docroot.  We try to avoid having the docroot be the repository root, but the build tool doesn't necessarily handle that nicely.

> Let's start with having that live under under the platform org, and if we
> run into any issues with that we can transfer it to Backdrop then.

Roger that.  I'll see what I can whip up in the next week or two.  Geoff, if you're around in our Slack channel I may bug you with questions. :smile:

**Comment by serundeputy** [comment]
yeah; i would like for the app to have a structure like:

```
config
  - active
  - staging
web
  - index.php
  - core
.platform.app.yml
.lando.yml
 ```

similar/same as we have now at https://github.com/platformsh/platformsh-example-backdrop my thoughts being that Backdrop users attracted to p.sh hosting will want to leverage configuration management and deploy workflows like:

```yml
    deploy: |
        # Import configuration on deploy.
        cd /app/web
        drush cc drush
        drush updb -y
        drush bcim -y
        drush cc all
```

**Comment by jenlampton** [comment]
Also maybe we can place private files outside the docroot?

**Comment by serundeputy** [comment]
>Also maybe we can place private files outside the docroot?

that seems fine @jenlampton is there anything we need to do like in `settings.php` @jenlampton to make backdrop aware of the location of the private files?

**Comment by jenlampton** [comment]
> is there anything we need to do like in settings.php

Not until https://github.com/backdrop/backdrop-issues/issues/4174 gets in, but I think we should add the directory so it's there if people need it.

**Comment by Crell** [comment]
Status report: 

WIP code is here: https://github.com/platformsh-templates/backdrop/pull/1
As produced by this build script: https://github.com/platformsh/template-builder/pull/162

And can be seen here: https://pr-1-djjnuwy-bf6tpupauqym2.eu-3.platformsh.site

There's still 3 outstanding issues:

* The override we use for Drupal to force the file, private, and temp paths to be what are set in Platform.sh doesn't work on Backdrop; @serundeputy and I tried to put an extra script in the deploy hook but it doesn't have access to the `config()` function.
* Maybe related, maybe not, while image uploads work image style generation is not working.  There are no errors in the log, just the files don't get generated.
* Most drush commands work, but `drush cron` does not.  Which... kinda makes it hard to configure cron. :-)  The error message is... well, now that I'm testing it again to see what the error is, now it's working.  I hate heisenbugs.  This requires some attention to see why it's intermittent.

Help on the first two points is most welcome, because I have no idea.

**Comment by Crell** [comment]
Following up here.  @jenlampton @serundeputy any idea how I can debug the image style generation, or handle the config setting?

**Comment by jenlampton** [comment]
> The override we use for Drupal to force the file, private, and temp paths to be what are set in Platform.sh doesn't work on Backdrop

There is a core PR to add these as overridable settings in `settings.php`. It might be easiest to apply the PR. The [PR is here](https://github.com/backdrop/backdrop/pull/2964) also [available as a patch](https://github.com/backdrop/backdrop/pull/2964.patch). See https://github.com/backdrop/backdrop-issues/issues/4174 for all the comments.

> Maybe related, maybe not, while image uploads work image style generation is not working. There are no errors in the log, just the files don't get generated.

Does platform use nginx? If so, an additional config override will need to be added to get image generation working in a sub-directory of files. I also wonder if this is related to the paths problem above?

**Comment by Crell** [comment]
Ahso.  Yes, we do use nginx, although we didn't need to change anything on Drupal to make that work.  What's the change needed?

I'll try the patch tomorrow.

**Comment by jenlampton** [comment]
> What's the change needed?

It's this one: https://forum.backdropcms.org/forum-topic/problems-image-styles

**Comment by Crell** [comment]
Ah ha!  OK, that forum thread tipped me off to a place where the `.platform.app.yaml` file was differently configured than in the Drupal 7 template. Once I copied over that part of the D7 template image styles worked.  Thanks.  (The passthru wasn't inheriting.)

I was able to apply the patch, and modify the settings file accordingly.  I *think* it's all working, but could you give the PR a once over visually to see if I'm doing anything obviously un-Backdrop-ish?

PR is here: https://github.com/platformsh-templates/backdrop/pull/1

It's committing all of Backdrop 1.14.1 so it's kinda big, but the key parts are the [.platform.app.yaml file](https://github.com/platformsh-templates/backdrop/pull/1/files#diff-fe14053ff51cb358f8b709415b2a2d17) and [`settings.platformsh.php`](https://github.com/platformsh-templates/backdrop/pull/1/files#diff-fd67b2001c1789aac2f19bc0726227ea) file.

Let me know if you see anything there that still needs to be changed.  If not, I think we can proceed with this code.

**Comment by jenlampton** [comment]
The PR looks good, the only thing I noticed won't cause a problem, but we don't have the global for `$drupal_hash_salt`. For us it's `$settings['hash_salt']`. Our `$settings` is is comparable to Drupal's `$conf`.

**Comment by Crell** [comment]
Thanks, Jen.  Code updated, PR merged.  It should appear in the new project wizard in the next day or three.

As far as ongoing maintenance, in theory the only maintenance needed will be periodically updating the version to download.  For that, if you file a PR against this file:

https://github.com/platformsh/template-builder/blob/master/project/backdrop.py#L7

And change the version number on that line, we can run a simple rebuild at that point.  It should be an easy process for whenever you have a new release.

I should have more information for you on the "Deploy on Platform" link/button later this week.

**Comment by jenlampton** [comment]
Thanks! I've updated our docs with the new process for each release.

**Comment by jenlampton** [comment]
oops, not sure if this should be closed...  reopening :)

**Comment by Crell** [comment]
:smile: Let's leave this open until I have the DoP button info for you.  Once that link is in place somewhere we can call this fixed.

**Comment by Crell** [comment]
Here's the markup to use.  It's just a link wrapped around an image tag, served from our site.  It points to the template definition file we have so users will get the exact same result as if they go through our wizard directly.  The utm tags are just so we know which links come from where so we know how much Backdrop traffic we get.

The image is an SVG is if you want to size it differently, go for it.  This is just the size we're using on GitHub.

```
<a href="https://console.platform.sh/projects/create-project?template=https://raw.githubusercontent.com/platformsh/template-builder/master/templates/backdrop/.platform.template.yaml&utm_content=backdrop&utm_source=backdrop&utm_medium=button&utm_campaign=deploy_on_platform">
    <img src="https://platform.sh/images/deploy/lg-blue.svg" alt="Deploy on Platform.sh" width="180px" />
</a>
```

**Comment by jenlampton** [comment]
I took the URL from the `href` and added it to the link on https://backdropcms.org/try-backdrop. Please let us know if you'd like any changes to the copy there. (That page is still a WIP.)

**Comment by Crell** [comment]
Text looks good to me.  The "trial setup" link bypasses the template, so anyone going through that link will get the normal checkout process with a list of templates to pick rather than going straight to Backdrop.  You may want to use the same link for both so they get Backdrop either way.

**Comment by jenlampton** [comment]
Done! :)

**Comment by Crell** [comment]
Looks like the link got cut off part way through when you copied it. :smile:

**Comment by jenlampton** [comment]
Fixed the link.
